add YOLO related tools for image segmentation

sunyi000 opened the pull request:
@bgruening just adding some yolo tools for building a workflow in Galaxy.
could you have a look? thank you very much

sunyi000 commented:
- I see the linting error is about `.shed.yml`, how can I fix that?
- should I move all test data files greater than 1M to somewhere and use the `location` attribute for `tests`?
Thanks

bgruening commented:
Yes please use location attribute and maybe put test files on zenido

bgruening commented:
For the shed tools warnings see:

.. WARNING: Repository suite definition repositories should have names starting with suite_

bgruening commented:
Once I'm back on a real notenook I will review, thanks @sunyi000

sunyi000 commented:
> Once I'm back on a real notenook I will review, thanks @sunyi000

No worries, have a good weekend. thank you. It's all green now..

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,154 @@
+	<section name="training_params" title="Training Parameters">
+            <param name="num_train" type="integer" value="70" label="How do you want to split your images for training?" help="Proportion of training images wrt valid and test images" />
+	    <param name="num_class" type="integer" label="Number of class names for images" />
+	    <param name="model_name" type="text">
```
You want here a `text` parameter? If the model name is not valid the tool will crash. What about making it a simple select include the most prominent models but offer an option to provide a file from the history?

sunyi000 replied:
I agree. I did it this way all because  the existing python takes this model_name text arg.

I think maybe better just leave the model_url select param, remove model_name, and then in <command> section find the name string...

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,154 @@
+    </configfiles>
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image(same name different extension)." />
```
```suggestion
        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image (same name different extension)." />
```

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,154 @@
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image(same name different extension)." />
+	<param name="ds_name" type="text" label="Name of dataset" help="The name for your dataset"/>
```
What is this parameter doing?

sunyi000 replied:
It is required in yolo.yaml which is created in python and needed during training

bgruening replied:
Can we hordcode something? Will the user ever see this? If it is really needed from a user I think we need to change the label or help section to describe what it is, especially because we deal with many files/datasets here.

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,154 @@
+                <param name="weight_decay" value="0.001" />
+            </section>
+	    <output_collection name="output_files" type="list" count="4">
+		<element name="results_metrics">
```
how many outputs are in here? Should we separate the text files from the plots? Maybe 4 different outputs, than everything in one collection?

sunyi000 replied:
yes 4 outputs for now.

I also think 4 different outputs may be better.

there are also other intermediate images as result of training, a lot, but for us they are not needed. I'm just thinking if in future other users want to tweak the output to include the intermediate images, a collection maybe easier.

bgruening commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,102 @@
+                        <has_size value="9615" delta="100" />
+                    </assert_contents>
+                </element>
+		<element name="pred-test01_mask" ftype="tif">
```
```suggestion
		<element name="pred-test01_mask" ftype="tiff">
```

bgruening commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,102 @@
+	<param name="max_det" type="integer" value="300" label="Max. of detections" help="Maximum number of detections allowed per image. Limits the total number of objects the model can detect in a single inference, preventing excessive outputs in dense scenes." />
+    </inputs>
+    <outputs>
+	 <collection name="results" type="list" label="YOLO prediction results">
```
I think it would be better to have two separate outputs and not one collection with two heterogenous elements.

sunyi000 replied:
just checked here, the no. of prediction result files depends.
for example, 
if the no. of images to predict is 10, the result no. of output files will be 20 (10masks, 10 txt)
maybe better to have collection here..

for the tests in this yolo_prediction.xml, image to predict is just one, that's why it only has 2 outputs.

sunyi000 commented on tools/image_processing/yolo/json2yolosegment.xml:
1. 2.I think Yes. Anylabeling is optional, if users already have run anylabeling somewhere else, they will already have the json and class name files.then the galaxy workflow will start without anylabeling.

3.anylabeling is just the annotation , I don't think yolo output can be integrated. Also, I think maybe better to keep them separate,so users can run either tool alone..

sunyi000 commented on tools/image_processing/yolo/adjust_save_images.xml:
I will test the other tool, if it already exists and works, then we could remove this tool.

sunyi000 commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,102 @@
+    </command>
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="num_class" type="integer" label="Number of class names for images" />
+	<param name="class_name" type="data" format="txt" label="YOLO class name file" />
```
I need to double check this one.

sunyi000 commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,102 @@
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="num_class" type="integer" label="Number of class names for images" />
+	<param name="class_name" type="data" format="txt" label="YOLO class name file" />
+	<param name="model" type="data" format="data" label="Model file" />
```
It's a pytorch model file, .pt file

bgruening replied:
Are those just the weights? We need to find a secure way here: https://github.com/pytorch/pytorch/security/policy#using-pytorch-securely

@anuprulez can maybe also help us here.

anuprulez replied:
I agree. Yolo checks if the uploaded `.pt` file is a valid Pytorch model or not in a proper way.

https://github.com/ultralytics/ultralytics/blob/da98efc61d9e0467315fc86c2297c8d81e656b1a/ultralytics/engine/model.py#L383

sunyi000 commented:
> Very cool, sorry for my naive questions, this is all relatively new to me! Thanks a lot @sunyi000 ...
> 
> * You can add a creator tag to all your tools and reference yourself, but maybe also NFDI?
> * citations should include YOLO I assume?
> * I think we have migrated everything to `tiff`, please do not use `tif` anymore
> * it would be great if we can add a min/max to all the integers/floats if possible

Thanks very much for the review. I will have a closer look next week.
I will address these 3 points.

The overview of our existing workflow 

1. Adjust save images (optional)
2. Anylabeling (optional)
3. Convert anylabeling json to yolo txt
4. train
5. predict.

I thought it would be cool to have anylabeling so we will have a human in the loop Galaxy workflow.

I will continue next week.

bgruening commented:
> I thought it would be cool to have anylabeling so we will have a human in the loop Galaxy workflow.

!!! This will be a super cool use-case.

kostrykin commented on tools/image_processing/yolo/json2yolosegment.xml:
```diff
@@ -0,0 +1,44 @@
+<tool id="json2yolosegment" name="JSON to YOLO text" version="@VERSION@+galaxy0" profile="24.2">
```
> the yolo training tool takes yolo text as input, and can run alone
I guess the annotation json, could be created by other software too, not only Anylabeling

yes, but I think the question that it boils down to is, is there any other tool that might require the YOLO text produced by this tool? If not, and only purpose of this tool is to create an input for the YOLO training tool, I think it would be more user-friendly to merge it into the YOLO training tool, no?

> also i think the code of yolo_train already gets big, adding the conversion functionality to the yolo wrapper may make the yolo_train code bloated..as we will need to consider all the possible annotation formats

then maybe we should rather think about creating a more general tool for conversion of vector-based annotations? Like we already have the bioformats converion tool, but for vector-based annotations?

kostrykin replied:
I would agree with you, if YOLO TXT would be some sort of a _standard_, but it isn't right? It's just for this one tool… Nevertheless, the decision is up to you, of course!

kostrykin replied:
You're right! Didn't know that.

kostrykin commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,102 @@
+<tool id="yolo_predict" name="YOLO Prediction" version="@VERSION@+galaxy0" profile="24.2">
+    <description>Prediction with YOLO models</description>
```
Yes but what is the name of the package that is wrapped here? Is it YOLO, YOLO8, or YOLO11? Judging by the versioning scheme I suspect looking at https://github.com/ultralytics/ultralytics/releases I'd suspect that the _name_ of the package is YOLO11 and the _version_ is 8.3.x?

sunyi000 replied:
we are not using a particular yolo model version here. User train their own model based on existing yolo models.
the wrapper python takes an argument 'model_name', which could be any, yolov8, yolov11-seg, sam etc..

sunyi000 commented on tools/image_processing/yolo/macros.xml:
```diff
@@ -0,0 +1,13 @@
+<macros>
+    <token name="@VERSION@">0.1.0</token>
```
i don't know. I thought this is a galaxy tool version, not the package or software version?

anuprulez commented on tools/image_processing/yolo/adjust_save_images.xml:
```diff
@@ -0,0 +1,85 @@
+     </inputs>
+    <outputs>
+        <collection name="output_jpgs" type="list" label="Adjusted JPEGS">
+		<discover_datasets pattern="(?P&lt;name&gt;.+)\.jpg" visible="true" directory="./output/jpegs" format="jpg"/>
```
```suggestion
	<discover_datasets pattern="(?P&lt;name&gt;.+)\.jpg" visible="true" directory="./output/jpegs" format="jpg"/>
```

anuprulez commented on tools/image_processing/yolo/adjust_save_images.xml:
```diff
@@ -0,0 +1,85 @@
+<tool id="adjust_save_images" name="Adjust input images for YOLO" version="@VERSION@+galaxy0" profile="24.2">
+    <description>Adjust contrast of input TIF images,and/or covnert them into 8bit RGB</description>
```
```suggestion
    <description>Adjust contrast of input TIF images and/or convert them into 8-bit RGB</description>
```

anuprulez commented on tools/image_processing/yolo/adjust_save_images.xml:
```diff
@@ -0,0 +1,85 @@
+        - A dataset collection of .jpg files.
+
+        Options:
+        - CLAHE: Contrast limited adjustment.
```
```suggestion
        - CLAHE: Contrast Limited Adaptive Histogram Equalization.
```

sunyi000 commented:
Hi all @bgruening @anuprulez @kostrykin ,

thanks very much for reviewing this.

could you please have another round of review? I believe I have addressed the comments.

@bgruening I have seen CI output in this commit, which I don't know how to fix, could you please help?

`FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/.planemo/gx_venv_3.11_release_24.1/bin/galaxyctl'`

- for the adjust_save_image tool, I removed the `clahe` part, since it already exists in another tool.
- but i left the convert to 8-bit part, I couldn't find existing tool does this, and at least for our existing workflow, it needs it.

- can we possibly move the yolo_training tool to run on GPU? I tested it locally with the below commands, and it is pretty fast, but on CPU, it takes a while

`planemo t --docker --docker_run_extra_arguments='--gpus all --shm-size=8g'`

I hope i didn't miss anything, but if I do, please let me know, it's lots of changes

Thank you very much

anuprulez commented on tools/image_processing/yolo/adjust_save_images.py:
```diff
@@ -0,0 +1,46 @@
+from skimage.io import imread, imsave
+
+
+def apply_clahe(image):
```
`apply_clahe` is not used here?

sunyi000 replied:
The function already exists in other existing tool, so I removed the function

sunyi000 replied:
not referenced. I will remove it. Thanks

anuprulez commented on tools/image_processing/yolo/preprocessing.py:
```diff
@@ -0,0 +1,73 @@
+        f.write("test: test\n")
+        f.write("\n")
+        f.write(f"nc: {meta['training_params']['num_class']}\n")
+        f.write("names: ['dataset']\n")
```
the `names` should always contain `dataset` keyword?

sunyi000 replied:
here https://github.com/bgruening/galaxytools/pull/1634#discussion_r2133878510

bgruening commented on tools/image_processing/yolo/json2yolosegment.xml:
```diff
@@ -0,0 +1,49 @@
+      mkdir ./input ./output &&
+		    
+      #for $i, $filename in enumerate($in_json):
+          ln -s '$filename' './input/${filename.element_identifier}' &&
```
you are not using `$i` here, isn't it? You might want to use i to make sure the names are unique?

bgruening commented on tools/image_processing/yolo/json2yolosegment.xml:
```diff
@@ -0,0 +1,49 @@
+         <param name="class_name" type="data" format="txt" label="Class file" help="Class names text file, contains the names of classes, one class per row"/>
+    </inputs>
+    <outputs>
+        <collection name="output_yolo" type="list" label="YOLO Text File">
```
```suggestion
        <collection name="output_yolo" type="list" label="YOLO Text Files">
```

bgruening commented on tools/image_processing/yolo/json2yolosegment.xml:
```diff
@@ -0,0 +1,49 @@
+            <param name="class_name" value="class_names.txt" />
+            <output_collection name="output_yolo">
+                <element name="in_json">
+                    <assert_contents><has_size value="1819" /></assert_contents>
```
please do not use `size`. This is not a good test. Try to assert the content. How many lines, rows, text snippets that are included ...

bgruening commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,100 @@
+    </macros>
+    <expand macro="creator" />
+    <requirements>
+	<container type="docker" >quay.io/sunyi000/ultralytics-galaxy:latest</container>
```
```suggestion
	<container type="docker">quay.io/sunyi000/ultralytics-galaxy:latest</container>
```

bgruening replied:
Where is this Dockerfile? Should we put it into a simliar setup as the IT? With CI and so on?

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,148 @@
+    </macros>
+    <expand macro="creator" />
+    <requirements>
+	<container type="docker" >quay.io/sunyi000/ultralytics-galaxy:latest</container>
```
should this be moved to the macro?

And it would be nice to give it a version that is somehow indicative

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,148 @@
+      mkdir -p ./input_images ./input_yolo ./training ./runs ./models output_files && 
+
+      #for $i, $filename in enumerate($input_images):
+          ln -s '$filename' './input_images/${filename.element_identifier}' &&
```
$i not used?

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,148 @@
+      #end for
+	
+      #for $s, $yolo in enumerate($input_yolo):
+          ln -s '$yolo' './input_yolo/${yolo.element_identifier}' &&
```
$s not used?

bgruening replied:
Do the names need to match between input_yolos and input_images?

sunyi000 replied:
yes the names must match.

If input_images and input_yolo are from anylabeling, they definitely match, anylabeling ensures this.

anylabeling outputs json files which have the same file name as the images.

then the json2yolosegment.xml also ensures the output txt, the input_yolo has the same file name as the input json.

Outside a workflow, if users want to use yolo_train alone, then it's users responsibility to make sure these names match.

bgruening replied:
I think that is fine. Maybe mention that in the tool help?

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,148 @@
+          ln -s '$yolo' './input_yolo/${yolo.element_identifier}' &&
+      #end for
+
+      python '$__tool_directory__/preprocessing.py' -i './input_images' -y './input_yolo' -o ./training -p '$num_train' -m $input_json &&
```
```suggestion
      python '$__tool_directory__/preprocessing.py' -i './input_images' -y './input_yolo' -o ./training -p '$num_train' -m '$input_json' &&
```

bgruening commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,148 @@
+	    </output>
+            <output name="metrics_plot">
+                <assert_contents>
+                    <has_size value="344784" delta="100"/>
```
There are some better image assertions now that you could reuse.

sunyi000 commented:
> > > ```
> > > * but i left the convert to 8-bit part, I couldn't find existing tool does this, and at least for our existing workflow, it needs it.
> > > ```
> > 
> > 
> > @kostrykin do we need a general tool for this to help Yi?
> 
> Probably I'm missing something here—please correct me!—but why can't we do the conversion to uint8 inside the training/prediction wrappers?
> 
> > could you please have another round of review? I believe I have addressed the comments.
> 
> Thanks @sunyi000! I will have a look in the next days.


We have some images which caused some weird shape errors for yolo train, so we converted those images to 8-bit jpeg. That's the only reason we created this adjust_save_image tool.

The yolo train tool can take images which are not 8-bit, so include the conversion in yolo train script will be too specific to one use case, I think.

We can completely exclude this convertion/adjust tool from the workflow if we know the input images can be used with yolo.

bgruening commented:
@sunyi000 I think I fixed the repo and the pushing of the container

kostrykin reviewed (commented):
@sunyi000 Can you please clarify, do you want to indicate that

**(a) these wrappers are for *ultralytics***, in terms of
- *"the Galaxy tool is a wrapper for an underlying tool"* in the [IUC recommendations](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions) and that
- the *"Galaxy tool is a thin tool wrapper"* in the [Naming and Annotation Conventions for Tools in the Image Community in Galaxy](https://doi.org/10.37044/osf.io/w8dsz), or

**(b) these wrappers are *not* for *ultralytics***, in terms of
- *"the Galaxy tool cannot be identified with a single underlying tool"* in the [IUC recommendations](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions) and that
- *"the main functionality of the tool is implemented in the wrapper"* itself in the [Naming and Annotation Conventions for Tools in the Image Community in Galaxy](https://doi.org/10.37044/osf.io/w8dsz)?

Because this decides the path that we need to take to properly annotate the tools. I think this also is an important decision and distinction to make in terms of FAIR'ness and consistency. 

I saw that you accepted https://github.com/bgruening/galaxytools/pull/1634#discussion_r2137347153 which indicates **case (a)**, but you also dismissed https://github.com/bgruening/galaxytools/pull/1634#discussion_r2137311284 and this dismissal rather indicates **case (b)**. I think we need to be consistent here.

sunyi000 commented on tools/image_processing/yolo/.shed.yml:
```diff
@@ -0,0 +1,17 @@
+categories:
+  - Imaging
+description: tools for building yolo workflows.
+long_description:   a collection of tools for training yolo model.
```
thanks, i will just manually make the changes, and trigger 1 CI job in the end once I finish everything.

bgruening commented:
> > @sunyi000 I think I fixed the repo and the pushing of the container
> 
> I got error `unauthorized: access to the requested resource is not authorized` open the quay.io repo also shows an unauthorized error..
> 
> Do i need to re-create the release?
> 
> @bgruening the quay.io repo is working now. Danke schön

Is this still an issue? I can open https://quay.io/repository/bgruening/docker-ultralytics?tab=settings ...

sunyi000 commented:
> > @bgruening the quay.io repo is working now. Danke schön
> 
> Is this still an issue? I can open https://quay.io/repository/bgruening/docker-ultralytics?tab=settings ...

not an issue any more, thanks

kostrykin commented on tools/image_processing/yolo/json2yolosegment.xml:
```diff
@@ -0,0 +1,56 @@
+<tool id="json2yolosegment" name="Convert AnyLabeling JSON to YOLO text" version="@VERSION@+galaxy0" profile="24.2">
```
Minor change to comply with IUC:
```suggestion
<tool id="json2yolosegment" name="Convert AnyLabeling JSON to YOLO text" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="24.2">
```

kostrykin reviewed (commented):
Much thanks @sunyi000! A few comments inside.

In addition:

**1)** Just so we do not forget about, this still needs to be addressed: https://github.com/bgruening/galaxytools/pull/1634/files#r2142215422

**2)** I think you should also add to the help section, that only 2-D images are supported. What about multi-channel images, like RGB images? If not supported, please explicitly mention this in the help section.

**3)** In this case, if this makes sense, please point to the tools that can be used to convert a multi-channel image to a list of single-channel images (like [**Split image along axes** with NumPy](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/split_image/ip_split_image), for example), so that YOLO tools still can be used to process those images (by the way, the same tool can also be used to process images of higher order than 2-D).

**4)** Do we need to account for https://docs.pytorch.org/docs/stable/notes/randomness.html here, like we did in https://github.com/bgruening/galaxytools/pull/1593? Or is this already taken care of in the wrapper? I think @bgruening mentioned something similar before, but I couldn't find it, but thought it's better to ask twice than to miss it…

kostrykin commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,107 @@
+              <option value="segment">segment</option>
+              <option value="detect">detect</option>
+	</param>
+	<param name="image_size" type="integer" value="1000" min="16" max="2048" label="Image size" help="Size of input image to be used only as integer of w,h." />
```
I think there should be no upper limit for `image_size`, or have I missed something in the docs? I think it's also good to add a brief comment on what this parameter does:
```suggestion
	<param name="image_size" type="integer" value="1000" min="16" label="Image size" help="All input images will be re-sized to squares with sides of this length (in pixels). This value governs the trade-offs of speed (smaller values) vs accuracy (larger values)." />
```

sunyi000 replied:
i guess this is galaxy xml recommendation, that all integer float param should have a min, max value.

it can be any value i think

kostrykin commented on tools/image_processing/yolo/json2yolosegment.py:
```diff
@@ -0,0 +1,64 @@
+                    print(f"No points found for annotation '{label}', skipping.")
+                    continue
+
+                x = [point[0] / image_width for point in points]
+                y = [point[1] / image_height for point in points]
```
I couldn't find documentation on the format, but from the code, I suspect that `point` is in pixel coordinates, whereas `x` and `y` need to be between `0` and `1`. In that case, there probably is a small bug here:
```suggestion
                x = [point[0] / (image_width - 1) for point in points]
                y = [point[1] / (image_height - 1) for point in points]
```
This is because, for points that are in the last row/column of the image, we want to have coordinates of 1. With the original code, this wouldn't be the case due to zero-based numbering, right?

sunyi000 replied:
i don't quite understand it. this is just reading from anylabeling generated json..points, image_width/height is just reading whatever output from anylabeling, so this should be already looked after

kostrykin replied:
Maybe I'm the one who misunderstands something :) lets brake it down:
**A1.** `image_width` is the width of the image, that comes from AnyLabeling?
**A2.** `point[0]` is the horizontal coordinate of a point, in pixels, that comes from AnyLabeling?
**A3.** For YOLO TXT, `x` needs to be in [0,1], where 0 is the left-most pixel and 1 is the right-most pixel of the image?

If the above is true, then, lets see what we get with the original code:
- Suppose `image_width` is 10
- Suppose `point[0]` is 0 (corresponds to a left-most pixel)
- Then we get `x = 0` which is fine ✅
- Now, suppose `point[0]` is 9 (corresponds to the right-most pixel)
- Then we get `x = 0.9` which contradicts **A3** from above. ❌

With the fixed code, and `image_width = 10` as an example, we get:
- `x = 0` if `point[0]` is 0, i.e. a left-most pixel ✅
- `x = 1` if `point[0]` is 9, i.e. a right-most pixel ✅

sunyi000 replied:
sorry i still don't understand. 
A1,2,3 -> yes
how is x=0.9 contradicts A3?
sorry the code is written by one of our image anylysts, i would have to ask him.

sunyi000 replied:
ok. understood. Thanks a lot. i will change the code

kostrykin commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,149 @@
+	 <inputs name="input_json"/>
+    </configfiles>
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
```
> > **2)** I think you should also add to the help section, that only 2-D images are supported. What about multi-channel images, like RGB images? If not supported, please explicitly mention this in the help section.
> 
>  […]
> * 2,3.  I couldn't find any docs explicitly says this, but it seems all yolo models are trained using jpeg.

The tool accepts `png` and `tiff` besides of `jpg`, but the tests only cover the `jpg` case. Can you maybe add a (small) test for `png` and `tiff`, just so that we can be sure that the wrapper does handle conversion properly?

Or, alternatively:
```suggestion
        <param name="input_images" type="data" format="jpg" multiple="true"  label="Input images"/>
```

kostrykin commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -0,0 +1,107 @@
+    ]]>
+    </command>
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
```
> > **2)** I think you should also add to the help section, that only 2-D images are supported. What about multi-channel images, like RGB images? If not supported, please explicitly mention this in the help section.
> 
>  […]
> * 2,3.  I couldn't find any docs explicitly says this, but it seems all yolo models are trained using jpeg.

The tool accepts `png` and `tiff` besides of `jpg`, but the tests only cover the `jpg` case. Can you maybe add a (small) test for `png` and `tiff`, just so that we can be sure that the wrapper does handle conversion properly?

Or, alternatively:
```suggestion
        <param name="input_images" type="data" format="jpg" multiple="true"  label="Input images"/>
```

kostrykin replied:
> it cannot be only jpg. it works with tiff, haven't test png.
will extra test really neccessary?

Sorry we found recently that many tools claim to support certain file types which they do not, which creates negative user experiences, this is why I'm being a bit nitpicky here, maybe.

> that require another zenodo collections of images, it won't be small because it's for training, and i would say minimum we need 10 images each format, because we need to split.

I understand the issue for training, but for prediction it should be fine to just add a tiny test file for TIFF, and another tiny test file for PNG, right?

sunyi000 replied:
ok. i will add tests for prediction and remove png tiff from training.xml

sunyi000 replied:
@kostrykin done. all suggested changes should be include, except for the additional section for augmentation. I would prefer leave it as-is for now..maybe until the code supports other tasks, such as `tracking` etc..
thanks a lot for the review

kostrykin reviewed (commented):
Thanks @sunyi000!

sunyi000 commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,149 @@
+    </configfiles>
+    <inputs>
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image (same name different extension)." />
```
see here https://github.com/bgruening/galaxytools/pull/1634#discussion_r2141074726

sunyi000 commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,149 @@
+        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image (same name different extension)." />
+	<section name="training_params" title="Training Parameters">
+            <param name="num_train" type="integer" value="70" min="50" max="90" label="How do you want to split your images for training?" help="Proportion of training images wrt valid and test images" />
+	    <param name="num_class" type="integer" value="1" min="1" max="10" label="Number of class names for images" />
```
it's the number of different object in the image that need to segment

same, it's just xml recommendation that all interger, float should have min, max, it can be any number. which number would be better in this case?

kostrykin replied:
Yes `integer` is fine here, why do you want to change it to `float`, am I missing something?

What I'm actually trying to say is that **this field can probably be removed entirely**, this is what the maintainer wrote on https://github.com/ultralytics/ultralytics/issues/13789#issuecomment-2626359322:
> For clarity, the nc parameter is actually optional in YAML files as class counts are automatically inferred from the names list.

sunyi000 replied:
i will test without it, if it works i will remove it. thanks

sunyi000 commented on tools/image_processing/yolo/yolo_training.xml:
```diff
@@ -0,0 +1,149 @@
+        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
+        <param name="input_yolo" type="data" format="txt" multiple="true"  label="Input YOLO txt files" help="The YOLO text files, each text file must correspond to one input image (same name different extension)." />
+	<section name="training_params" title="Training Parameters">
+            <param name="num_train" type="integer" value="70" min="50" max="90" label="How do you want to split your images for training?" help="Proportion of training images wrt valid and test images" />
```
i would say add a hint in help, because changing it to float require some modifications to the python code, which I'm a bit hesitate to do ..:)

kostrykin commented on tools/image_processing/yolo/yolo_prediction.xml:
```diff
@@ -38,7 +38,7 @@
     ]]>
     </command>
     <inputs>
-        <param name="input_images" type="data" format="jpg,png,tiff" multiple="true"  label="Input images"/>
```
> for prediction, i added test for png. it didnot work for tiff..so currently input only supports jpeg, png

Cool that you did the test :)

kostrykin reviewed (approved):
Much thanks @sunyi000!

I think we're good to merge now.

I noticed that some of your `help="…"` texts have a period `.` at the end, while others have not. You might want to make this consistent, if you want and have time (personally I think it's good to have one at the end of the `help="…"` text).

Let me know when you think we can merge.

yvanlebras commented:
AAmmaaazzzingg ! Looking forward to test it for biodiversity images !!!
